### PR TITLE
Sharding, Cache, Types, ESLint Comments

### PR DESCRIPTION
**Pull request by Digital39999**

### CODE REVIEW

File: src/events/interactionCreate.ts [hunk 1]
```diff
@@ -1,61 +1,20 @@
...
 
 export default {
     name: 'interactionCreate',
     run: async (interaction: CommandInteraction | ButtonInteraction | ContextMenuCommandInteraction) => {
-        const { client } = interaction;
-        if (!client.isReady()) return;
+        if (!interaction.client.isReady()) return;
 
         stats.num++;
-        console.log(`\x1b[2m\x1b[43m\x1b[37m\x1b[1m\x1b[37m${new Date().toISOString().replace('T', ' ').replace('Z', '')}:\x1b[0m \x1b[33m\x1b[1m${stats.num} :: ${ms(client.uptime ?? 0)}\x1b[0m`);
```
Comment (shi-gg) [inline src/events/interactionCreate.ts:15]: Why did you remove that lol

File: src/events/ready.ts [hunk 2]
```diff
@@ -1,33 +1,42 @@
...
+                description: command.description,
+                options: command.options,
+                dm_permission: command.dm_permission,
+            });
         });
 
-        if (process.platform === 'win32') return;
+        client.application?.commands.set(Interactions);
+
+        if (Config.dlist) {
```
Comment (shi-gg) [inline src/events/ready.ts]: Why `Config.dlist` if there is the `postStats(client);` few lines beneath it which isn't just dlist. Also why even that if check if it's required inside of the [config.ts](src/example.config.ts), see [typings.ts](src/typings.ts)

File: src/example.config.ts [hunk 3]
```diff
@@ -20,6 +21,8 @@ export const Config: ConfigType = {
         enabled: false,
     },
 
+    shards: 2,
```
Comment (shi-gg) [inline src/example.config.ts]: This should default set to `'auto'` so it always uses the recommended shard count from [`/gateway/bot`](https://discord.com/developers/docs/topics/gateway#get-gateway-bot)

File: src/handlers/events.ts [hunk 4]
```diff
@@ -1,13 +1,15 @@
-import { Client } from 'discord.js';
 import { readdirSync } from 'node:fs';
+import { Client } from 'discord.js';
 import { Event } from '../typings';
+import path from 'node:path';
 
 export default async function (client: Client) {
-    const eventFiles = readdirSync('./dist/events').filter((file) => file.endsWith('.js'));
+    const eventFiles = readdirSync(path.join('..', 'dist', 'events')).filter((file) => file.endsWith('.js'));
```
Comment (shi-gg) [inline src/handlers/events.ts:8]: Why do you use `path.join()` here but not two lines beneath?

File: src/handlers/interactions.ts [hunk 5]
```diff
@@ -1,11 +1,12 @@
 import { readdirSync } from 'node:fs';
-import { Config, Emote } from '../config';
+import { Config } from '../config';
+import path from 'node:path';
 
-export default function () {
-    const commands = readdirSync('./dist/interactions/commands/').filter(file => file.endsWith('.js'));
+export default async function () {
+    const commands = readdirSync(path.join('..', 'dist', 'interactions', 'commands')).filter((file) => file.endsWith('.js'));
```
Comment (shi-gg) [inline src/handlers/interactions.ts:7]: Why do you use `path.join()` here but not two lines beneath?

File: src/interactions/commands/mp3.ts [hunk 6]
```diff
@@ -1,27 +1,56 @@
...
 
         await interaction.deferReply({ ephemeral: !interaction.guild?.members.me?.permissionsIn(interaction.channelId).has(['ViewChannel', 'SendMessages', 'AttachFiles']) || visibility === 'hidden' });
 
-        // @ts-expect-error I dont understand those djs typings
-        const textInput: string = interaction.options.getString('text');
-        // @ts-expect-error I dont understand those djs typings
-        const speaker: string = interaction.options.getString('speaker');
-        const user = await users.findOne({ user: interaction.user.id });
+        const textInput: string = (interaction.options as CommandInteractionOptionResolver).getString('text') || 'No text provided';
+        const speaker: string = (interaction.options as CommandInteractionOptionResolver).getString('speaker') || 'en_us_002';
```
Comment (shi-gg) [inline src/interactions/commands/mp3.ts]: Redundant `string` typing

File: src/interactions/commands/voice.ts [hunk 7]
```diff
@@ -35,58 +52,54 @@ export default {
...
-        await interaction.deferReply();
-        // @ts-expect-error I dont understand those djs typings
-        const textInput: string = interaction.options.getString('text');
-        // @ts-expect-error I dont understand those djs typings
-        const speaker: string = interaction.options.getString('speaker');
-        const user = await users.findOne({ user: interaction.user.id });
+        await interaction.deferReply().catch(() => null);
 
+        const textInput: string = (interaction.options as CommandInteractionOptionResolver).getString('text') || '';
+        const speaker: string = (interaction.options as CommandInteractionOptionResolver).getString('speaker') || '';
```
Comment (shi-gg) [inline src/interactions/commands/voice.ts:59]: Redundant `string` typing

File: src/modules/getData.ts [hunk 8]
```diff
@@ -1,10 +1,14 @@
-import { Config, Emote } from '../config';
+import { Config } from '../config';
+import path from 'node:path';
 import fs from 'node:fs';
 
 export async function getData(text: string, voice: string): Promise<string | null> {
-    const name = `${process.cwd()}/${new Date().getTime()}.mp3`;
+    const fileName = randomCode(10);
+    const name = path.join(__dirname, '..', 'audio', fileName + '.mp3');
```
Comment (shi-gg) [inline src/modules/getData.ts:8]: where is the problem using `process.cwd()`, which gives the root of the process

File: src/modules/getData.ts [hunk 9]
```diff
@@ -1,10 +1,14 @@
-import { Config, Emote } from '../config';
+import { Config } from '../config';
+import path from 'node:path';
 import fs from 'node:fs';
 
 export async function getData(text: string, voice: string): Promise<string | null> {
-    const name = `${process.cwd()}/${new Date().getTime()}.mp3`;
```
Comment (shi-gg) [inline src/modules/getData.ts:5]: why not just using the current time?

File: src/sharding.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+            }
+        });
+    });
+
+    manager.spawn();
+} catch (e) {
+    console.log(e);
+}
+
+// ------------------------------- clean console
```
Comment (shi-gg) [inline src/sharding.ts]: For what is this needed? If you want to get rid of the `fetch()` experimental warning, use [this from app.ts](src/app.ts#45)

File: src/sharding.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+console.log(`\x1b[36mStarting ${Config.shards} shards.\n\x1b[0m`);
+
+try {
+    let count = 0;
+    const manager = new ShardingManager('./app.js', {
+        token: Config.token,
+        totalShards: Config.shards,
+    });
+
+    manager.on('shardCreate', (shard) => {
```
Comment (shi-gg) [inline src/sharding.ts:20]: If you are already adding sharding, you might also fix getting the guild count which are posted to botlists  [events/ready.ts](src/events/ready.ts#48)

File: package.json [hunk 12]
```diff
@@ -1,17 +1,16 @@
...
   "dependencies": {
     "@discordjs/voice": "^0.14.0",
     "discord.js": "^14.7.1",
     "dlist.js": "^0.4.2",
     "ffmpeg-static": "^5.0.2",
+    "get-audio-duration": "^3.1.1",
     "libsodium-wrappers": "^0.7.10",
-    "mongoose": "^6.6.2",
-    "mp3-duration": "^1.1.0",
-    "ms": "^2.1.3"
```
Comment (shi-gg) [inline package.json:14]: If you remove ms, please also remove `@types/ms` from the `devDependencies`

File: src/interactions/commands/mp3.ts [hunk 13]
```diff
@@ -31,12 +60,14 @@ export default {
 
         await interaction.editReply({
             content: `${Emote.success} Here's your audio file!\n${Config.ad}`,
-            files: [res]
+            files: [{
+                attachment: res,
+                name: 'audio.mp3'
```
Comment (shi-gg) [inline src/interactions/commands/mp3.ts]: I like time more because each time a user download's it, it won't append `(1)` to the file, plus if a user sorts the explorer after "name", it still will be sorted after creation date

File: src/interactions/commands/stop.ts [hunk 14]
```diff
@@ -27,8 +29,6 @@ export default {
 
         interaction.reply({
```
Comment (shi-gg) [inline src/interactions/commands/stop.ts]: If you defer the reply, you should edit here

File: src/interactions/commands/voice.ts [hunk 15]
```diff
@@ -1,18 +1,35 @@
...
+import { CommandInteractionOptionResolver } from 'discord.js';
+import getAudioDurationInSeconds from 'get-audio-duration';
+import { validate } from '../../modules/voteValidation';
 import { User, users } from '../../structures/user';
+import { getData } from '../../modules/getData';
+import { Config, Emote } from '../../config';
 import { Command } from '../../typings';
-import { connections, isPlaying, disconnect } from '../../app';
-import { validate } from '../../modules/voteValidation';
+import fs from 'node:fs';
```
Comment (shi-gg) [inline src/interactions/commands/voice.ts]: please sort those imports
```ts
// external packages

// space

// internal files

**Review by shi-gg — CHANGES_REQUESTED** [review]

File: src/events/interactionCreate.ts (hunk 1 above)
Comment (Digital39999) [reply src/events/interactionCreate.ts:15]: Why is it needed?

Comment (shi-gg) [reply src/events/interactionCreate.ts:15]: logging
